From 019c06292d87571165502b87eee11e4cf7f40b72 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Christian=20Str=C3=B8mme?= <christian.stromme@qt.io>
Date: Wed, 15 Oct 2025 15:38:37 +0200
Subject: [PATCH 428/553] Windows: Fix QWindowWindows::requestUpdate

It's very possible for the platform window to be released and recreated
for a QWindow. We therefore still need to verify that we both have
a platform window, and that it's the same window we triggered the
update for.

This change amends 31c38d5f766366931cd127f7323f9c26ac5de498

Fixes: QTBUG-141128
Change-Id: Id6e3fcbe3c6d2d5aa30b2af256a74de2fc9e5f9e
Reviewed-by: Tim Blechmann <tim.blechmann@qt.io>
(cherry picked from commit f3e029719b14bf2180d247afebaceea3607a323c)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/plugins/platforms/windows/qwindowswindow.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/plugins/platforms/windows/qwindowswindow.cpp b/src/plugins/platforms/windows/qwindowswindow.cpp
index 907cb388140..d4da7b77b1b 100644
--- a/src/plugins/platforms/windows/qwindowswindow.cpp
+++ b/src/plugins/platforms/windows/qwindowswindow.cpp
@@ -4024,9 +4024,11 @@ void QWindowsWindow::requestUpdate()
                     // request or we are waiting for the event loop to process
                     // the Posted event on the GUI thread.
                     if (m_vsyncUpdatePending.testAndSetAcquire(UpdateState::Requested, UpdateState::Posted)) {
-                        QMetaObject::invokeMethod(w, [w] {
+                        QWindowsWindow *oldSelf = this;
+                        QMetaObject::invokeMethod(w, [w, oldSelf] {
+                            // 'oldSelf' is only used for comparison, don't access it directly!
                             auto *self = static_cast<QWindowsWindow *>(w->handle());
-                            if (self) {
+                            if (self && self == oldSelf) {
                                 // The platform window is still alive
                                 self->m_vsyncUpdatePending.storeRelease(UpdateState::Ready);
                                 self->deliverUpdateRequest();
-- 
2.51.2

